include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -166,12 +167,22 @@ struct FFT<2,true, Scalar, device_capabilities>
         // Run a subgroup-sized FFT, then continue with bigger steps
         subgroup::FFT<true, Scalar, device_capabilities>::__call(lo, hi);
         
+        // special first workgroup-shuffle - only if workgroupsize > subgroupsize
+        if (_NBL_HLSL_WORKGROUP_SIZE_ ^ glsl::gl_SubgroupSize()) 
+        { 
+            exchangeValues<SharedMemoryAccessor, Scalar>(lo, hi, threadID, glsl::gl_SubgroupSize(), sharedmemAdaptor);
+        }
+
         // The bigger steps
         for (uint32_t stride = glsl::gl_SubgroupSize() << 1; stride < _NBL_HLSL_WORKGROUP_SIZE_; stride <<= 1)
+        {   
+            // If we enter this for loop, then the special first workgroup shuffle went through, so wait on that
+            sharedmemAdaptor.workgroupExecutionAndMemoryBarrier(); 
             FFT_loop<SharedMemoryAccessor>(stride, lo, hi, threadID, sharedmemAdaptor);
+        }
 
-        // special last iteration
-        if (_NBL_HLSL_WORKGROUP_SIZE_ > glsl::gl_SubgroupSize())
+        // special last iteration - only if workgroupsize > subgroupsize
+        if (_NBL_HLSL_WORKGROUP_SIZE_ ^ glsl::gl_SubgroupSize())
```

Comment:
you do realize you can put all this code under a single `NBL_HLSL_WORKGROUP_SIZE > SubgroupSize`  ?

Also whats with the bitwise XOR instead of a `>` ?

Reply:
you branch multiple times on same condition, move all under one if